CONTRIBUTING.rst

```diff
@@ -0,0 +1,177 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The basic WarpX workflow is:
+
+- Fork the main repo (or update it if you already created it).
+- Implement your changes and push them on a new branch ``<branch_name>`` on your fork.
+- Create a Pull Request from branch ``<branch_name>`` on your fork to branch `master` on the main WarpX repo.
+
+First, let us setup your local git repo. Make your own fork of the main (``upstream``) WarpX repo:
+on the `WarpX Github page <https://github.com/ECP-WarpX/WarpX>`_, press the fork button.
+Then, you can execute:
+
+.. code-block:: sh
+
+   # These 4 first lines are the same as for a standard WarpX install
+   mkdir warpx_directory
+   cd warpx_directory
+   git clone --branch master https://bitbucket.org/berkeleylab/picsar.git
+   git clone --branch development https://github.com/AMReX-Codes/amrex.git
+
+   # Clone your fork on your local computer. You can get this address on your fork's Github page.
+   git clone https://github.com/<myGithubUsername>/WarpX.git
+   cd warpx
+   # Keep track of the main WarpX repo, to remain up-to-date.
+   git remote add upstream https://github.com/ECP-WarpX/WarpX.git
+
+Now you are free to play with your fork (for additional information, you can visit the
+`Github fork help page <https://help.github.com/en/articles/fork-a-repo>`_).
+
+.. note::
+
+   You do not have to re-do the setup above every time.
+   nstead, in the future, you need to update the ``master`` branch on your fork with
+
+   .. code-block:: sh
+
+      git checkout master
+      git pull upstream master
+
+Make sure you are on WarpX ``master`` branch with
+
+.. code-block:: sh
+
+   git checkout master
+
+in the WarpX directory.
+
+Create a branch ``<branch_name>`` (the branch name should reflect the piece of code you want to add, like ``fix_spectral_solver``) with
+
+.. code-block:: sh
+
+   git checkout -b <branch_name>
+
+and do the coding you want.
+It is probably a good time to look at the `AMReX documentation <https://amrex-codes.github.io/amrex/docs_html/>`_ and at the `AMReX Doxygen reference <https://ccse.lbl.gov/pub/AMReX_Docs/index.html>`_.
+Add the files you work on to the git staging area with
+
+.. code-block:: sh
+
+   git add <file_I_created> <and_file_I_modified>
+
+Commit & push your changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Periodically commit your changes with
+
+.. code-block:: sh
+
+   git commit -m "This is a 50-char description to explain my work"
+
+The commit message (between quotation marks) is super important in order to follow the developments and identify bugs.
+
+For the moment, commits are on your local repo only.
+You can push them to your fork with
+
+.. code-block:: sh
+
+   git push -u origin <branch_name>
+
+If you want to synchronize your branch with the ``master`` branch (this is useful when the ``master`` branch is being modified while you are working on ``<branch_name>``), you can use
+
+.. code-block:: sh
+
+   git pull upstream master
```

Comment:
I might have done a mistake in my own repo structure, but I sometimes need the bottom line if I wish to merge to master into my branches. Right now I am not sure why.
```suggestion
   git pull upstream master
   git merge upstream/master
```
The latter line can be necessary, if merging doesn't happen automatically after you use the first line.

Reply:
`git pull` implies a merge unless `--ff-only` is specified. I would potentially add `git fetch` but that's also included in pull.

Reply:
Good point!